coop/rchain/key-value-store/Key.scala

```diff
@@ -0,0 +1,81 @@
+/*               __          __                                         *\
+**    __________/ /_  ____ _/_/___                                      **
+**   / ___/ ___/ __ \/ __ `/ / __ \     RChain API                      **
+**  / /  / /__/ / / / /_/ / / / / /     (c) http://rchain.coop          **
+** /_/   \___/_/ /_/\____/_/_/ /_/                                      **
+\*                                                                      */
+
+package KeyValueStore
+
+
+// Keys are like Prolog terms, where the predicate
+// name is all letters and each parameter is
+// all digits or all letters.
+
+class Key(key: String)
+{
+  private val _parsed = parse(key)
+  val _name = _parsed._1
+  val _params = _parsed._2
+  val _arity = _parsed._2.length
+  var _key = _name + "("
+  for (i <- 0 until _params.length-1)
+    _key += _params(i).param + ","
+  _key += _params(_params.length-1).param + ")"
+
+  def this(name: String, params: String*)
+  {
+    this(name + "(" + params.mkString(",") + ")")
+  }
+
+  def parse(key: String): (String,Array[Param])  = 
+  {
+    // need to check the form of key
+    val name = key.slice(0, key.indexOfSlice("(")).trim
+    if (name.length == 0 || name.length == key.length || !allLetters(name))
+      { throw new Exception(s"Key.Parse(): malformed (1): $key") }
+    val paramsOriginal =
+      key.slice(key.indexOfSlice("(")+1, key.length-1).split(",")
+    val params = new Array[Param](paramsOriginal.length)
+
+    var i = 0
+    for (p <- paramsOriginal)
+    {
+      val param = p.trim
+      if ((allLetters(param) || allDigits(param))
+          // disallow the same <letters> appearing more than once
+          && (allDigits(param) || !params.contains(param)))
+      {
+        // Why doesn't this work?
+        // params(i) = MakeParam(param)
```

Comment:
Capitalization?  The only definition of makeParam I can find has a lowercase 'm'.